feat(client): add SSH support on the UI

lorenzo-cavazzi opened the pull request:
Add SSH support in the UI

/deploy #persist #notest renku-core=feature/ssh-into-sessions renku-notebooks=feat-ssh-access extra-values=notebooks.ssh.enabled=true
fix #2324

ciyer commented on client/src/features/display/displaySlice.ts:
```diff
@@ -0,0 +1,62 @@
+  initialState,
+  reducers: {
+    showSshModal: (state, action: PayloadAction<ProjectConfig>) => {
+      return {
+        ...state, modals: {
+          ...state.modals, ssh: {
+            ...state.modals.ssh,
+            show: true,
+            projectPath: action.payload.projectPath,
+            gitUrl: action.payload.gitUrl,
+          }
```
This kind of thing isn't necessary. The redux-toolkit reducer arguments are [wrapped using Immer](https://redux-toolkit.js.org/usage/immer-reducers) so you can just mutate the values.

```suggestion
      state.modals.show = true;
      state.modals.projectPath = action.payload.projectPath;
      state.modals.gitUrl = action.payload.gitUrl;
```

lorenzo-cavazzi replied:
Oh nice! That simplifies reading the code. I'll add this and the following changes (I'll add another commit since I need to add `.ssh` and I'll use a single object to further simplify)

ciyer commented on client/src/features/display/displaySlice.ts:
```diff
@@ -0,0 +1,62 @@
+      };
+    },
+    hideSshModal: (state) => {
+      return { ...state, modals: { ...state.modals, ssh: { ...state.modals.ssh, show: false } } };
```
See above comment.

```suggestion
     state.modals.show = false;
```

ciyer commented on client/src/features/display/displaySlice.ts:
```diff
@@ -0,0 +1,62 @@
+      return { ...state, modals: { ...state.modals, ssh: { ...state.modals.ssh, show: false } } };
+    },
+    toggleSshModal: (state) => {
+      return { ...state, modals: { ...state.modals, ssh: { ...state.modals.ssh, show: !state.modals.ssh.show } } };
```
See above comment.

```suggestion
      state.modals.ssh.show: !state.modals.ssh.show;
```

ciyer commented on client/src/features/versions/versions.d.ts:
```diff
@@ -0,0 +1,44 @@
+interface NotebookComponent {
+  data: {
+    anonymousSessionsEnabled: boolean;
+    cloudstorageEnabled: {
```
I think it would be better of "Cloud Storage" were two word:

```suggestion
    cloudStorageEnabled: {
```

lorenzo-cavazzi replied:
I agree that `cloudStorageEnabled` would read better. I kept that aligned with the response from the service.

I would either keep this as is now, or change also `azure_blob` to `azure` or `azureBlob` (if we don't care about consistency with the response, then let's use camel case everywhere):

lorenzo-cavazzi replied:
Ah nevermind, let's stick to camelCase for coherence -- I'll change it

ciyer commented on client/src/features/versions/versions.d.ts:
```diff
@@ -0,0 +1,44 @@
+  version: string;
+  anonymousSessionsEnabled: boolean;
+  sshEnabled: boolean;
+  cloudstorageEnabled: boolean;
```
See above. 

```suggestion
  cloudStorageEnabled: boolean;
```

ciyer commented on client/src/styles/bootstrap_ext/_button.scss:
```diff
@@ -39,6 +39,10 @@ $borderRadius: 1000px;
   margin-right: 12px;
 }
 
+.btn-with-menu-options .btn-with-menu-margin {
+  margin-right: 12px;
+}
+
```
I'm guessing that the idea is that the margin should be the same for both `svg` and elements with the specified class. I think in that case, it would be better to combine the definitions:

```suggestion
.btn-with-menu-options {
    svg, .btn-with-menu-margin {
      margin-right: 12px;
     }
}
```
(Not 100% sure on that syntax, but should be something similar.)

lorenzo-cavazzi replied:
Ah yes, good point. I'll harmonize that CSS rules with the previous one on `svg` elements

ciyer reviewed (changes requested):
Still working on testing the UI, but I have reviewed the code. The main thing I noticed is that the reducers can be simplified. Otherwise, it looks good!

ciyer commented on client/src/utils/components/ssh/ssh.tsx:
```diff
@@ -0,0 +1,223 @@
+        <p>
+          To update your project, go to{" "}
+          <Link to={updateUrl} onClick={() => closeModal()}>Overview Status</Link>
+          {" "}then come back to this window to view the commands to connect remotely to sessions.
```
I think "come back to this window" is a bit misleading, since the window disappears when you go to the status page. Might better to say:

```suggestion
          {" "}start an ssh session again to view the commands to connect remotely to sessions.
```

lorenzo-cavazzi replied:
I further rephrased it based on both yours and other offline comments.
Please let me know if it works well

ciyer commented on client/src/utils/components/ssh/ssh.tsx:
```diff
@@ -0,0 +1,223 @@
+    modalBody = (
+      <>
+        <p>
+          <img src={rkIconSshTicked} className="rk-icon rk-icon-md filter-green me-2" />
+          Your project supports SSH.
```
I think this text here is very small / easy to overlook.

<img width="514" alt="image" src="https://user-images.githubusercontent.com/1196411/221536090-9313834e-0d1f-4e5b-a026-0db3c96eb2d0.png">

I think we should make it a bit more prominent and set it off from the rest of the body text somehow.

lorenzo-cavazzi replied:
Good point, I made it bold. Let me know if it works well with the new content.

ciyer commented on client/src/utils/components/ssh/ssh.tsx:
```diff
@@ -0,0 +1,223 @@
+          Your project supports SSH.
+        </p>
+        <p>
+          You can now switch to your local CLI and run the following command to set up the project
+          correctly. Then follow the instructions to connect.
```
Is it necessary to have the project checked-out locally? If so, we should mention that.

ciyer commented on client/src/utils/components/ssh/ssh.tsx:
```diff
@@ -0,0 +1,223 @@
+        </p>
+        <p>
+          You can now switch to your local CLI and run the following command to set up the project
+          correctly. Then follow the instructions to connect.
```
I do not see any instructions to connect. After running the specified commands, I only see:
```
$ renku session ssh-setup
Info: Generating keys
Info: Writing SSH config
```

lorenzo-cavazzi replied:
Good point. Based on this and the previous comment, the page has been reworked to contain more information. Please let me know if it's an improvement in the right direction.
P.S. I mixed in other offline suggestions too.

lorenzo-cavazzi commented:
Thank you for the review @ciyer :pray: 
I addressed all your points (mind that it takes ~40 mins to re-deploy due to referencing custom branches for renku-core and renku-notebooks).

Please double-check the changes in the text for the SSH instructions. Inspired by your and other offline comments, I ended up adding more information.

ciyer reviewed (approved):
This looks great! 🎉  As discussed in the stand up, I added a small commit with minor cosmetic changes for visual consistency.
